### PR TITLE
batches: add Bitbucket Cloud support

### PR DESCRIPTION
**Pull request by LawnGnome**
This PR implements the `ChangesetSource` and `ForkableChangesetSource` interfaces for Bitbucket Cloud.

In terms of functionality, there aren't any terribly unusual quirks here. The only really interesting one is that you can't reopen PRs on Bitbucket Cloud, but you can create multiple PRs for the same `(source, target)` combination, so the reopen operation is implemented as a wrapper for the create operation. It works out about the same in practice; the fact that the PR ID and other metadata changes is handled by `SetMetadata` anyway. 🤷 

In terms of the review, here are the things I'm going to call out as looking a little weird:

1. I converted `bitbucketcloud.Client` into an interface so it could be mocked in the source tests. There's a little noise associated with that.
2. Just over half the PR size is actually the generated mock. You can obviously ignore that. (Unless you don't want to, in which case… uh, thanks?)
3. There's a `TODO` in the event handling code, because we don't actually handle any events yet for Bitbucket Cloud. That's coming Real Soon Now™️ with webhook support, but isn't in this PR.
4. To support the reopen functionality described above, the executor now has to fill out the `Changeset` properly when invoking `ReopenChangeset`. That was probably a bug anyway.

## Test plan

A mixture of unit test coverage (which, if I do say so myself, is _excellent_), and actual, real world testing of the array of publication and bulk action scenarios that we support.

### CODE REVIEW

File: enterprise/internal/batches/sources/bitbucketcloud_test.go [hunk 1]
```diff
@@ -0,0 +1,889 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/bitbucketcloud"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+func TestNewBitbucketCloudSource(t *testing.T) {
+	t.Run("invalid", func(t *testing.T) {
+		for name, input := range map[string]string{
+			"invalid JSON":    "invalid JSON",
+			"unparsable JSON": `{"appPassword": ["not a string"]}`,
```
Comment (eseliger) [inline enterprise/internal/batches/sources/bitbucketcloud_test.go:28]:
```suggestion
			"invalid schema": `{"appPassword": ["not a string"]}`,
```

I'm sorry, super nitpicky. This is technically parseable JSON.

File: enterprise/internal/batches/sources/bitbucketcloud_test.go [hunk 2]
```diff
@@ -0,0 +1,889 @@
+package sources
```
Comment (eseliger) [inline enterprise/internal/batches/sources/bitbucketcloud_test.go:1]: nice test suite!

**Review by eseliger — APPROVED** [review]
Code-wise this all looks very good to me! Great job :)
I will do some dogfooding once we have the events parsing in, so I don't report things you are already aware of (if there are any). 
(Unless you want me to do it now??)

File: enterprise/internal/batches/sources/bitbucketcloud_test.go (hunk 1 above)
Comment (LawnGnome) [reply enterprise/internal/batches/sources/bitbucketcloud_test.go]: 48c

File: enterprise/internal/batches/sources/bitbucketcloud.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	// more details.)
+	//
+	// It will, however, allow a pull request to be recreated. So we're going to
+	// do something a bit different to the other external services, and just
+	// recreate the changeset wholesale.
+	//
+	// If the PR hasn't been declined, this will also work fine: Bitbucket will
+	// return the same PR in that case when we try to create it, so this is
+	// still (effectively) a no-op, as required by the interface.
+	_, err := s.CreateChangeset(ctx, cs)
```
[the remark this answers is not in the document]
Comment (LawnGnome) [reply enterprise/internal/batches/sources/bitbucketcloud.go:167]: We do! It's actually kind of weird watching the PR number change when you're viewing a batch change, but it works just fine.

File: enterprise/internal/batches/rewirer/rewirer_test.go [hunk 4]
```diff
@@ -26,7 +26,7 @@ func TestRewirer_Rewire(t *testing.T) {
 	unsupportedTestRepo := &types.Repo{
 		ID: unsupportedTestRepoID,
 		ExternalRepo: api.ExternalRepoSpec{
-			ServiceType: extsvc.TypeBitbucketCloud,
+			ServiceType: extsvc.TypeOther,
```
[the remark this answers is not in the document]
Comment (LawnGnome) [reply enterprise/internal/batches/rewirer/rewirer_test.go:29]: When I wrote the first iteration of this code, it was late and I was tired, and I was _so confused_ for about three minutes as to how this seemingly unrelated test was breaking.

**Comment by LawnGnome** [comment]
> (Unless you want me to do it now??)

I'm OK waiting — we're not shipping this until May anyway, and I expect the remaining PRs will be in next week.

**Pull request merged**